src/envoy/http/error_translator/filter.cc

```diff
@@ -0,0 +1,194 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/envoy/http/error_translator/filter.h"
+
+#include <string>
+
+#include "absl/types/optional.h"
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+#include "common/grpc/common.h"
+#include "common/grpc/status.h"
+#include "common/http/headers.h"
+#include "src/envoy/utils/filter_state_utils.h"
+#include "src/google/protobuf/util/json_util.h"
+
+namespace espv2 {
+namespace envoy {
+namespace http_filters {
+namespace error_translator {
+
+using ::Envoy::Http::FilterDataStatus;
+using ::Envoy::Http::FilterHeadersStatus;
+
+Filter::Filter(FilterConfigSharedPtr config) : config_(config) {
+  type_helper_ = std::make_unique<google::grpc::transcoding::TypeHelper>(
+      Envoy::Protobuf::util::NewTypeResolverForDescriptorPool(
+          Envoy::Grpc::Common::typeUrlPrefix(), &descriptor_pool_));
+
+  print_options_.add_whitespace = true;
+  print_options_.always_print_primitive_fields = true;
+  print_options_.always_print_enums_as_ints = false;
+  print_options_.preserve_proto_field_names = false;
+}
+
+FilterHeadersStatus Filter::encodeHeaders(
+    Envoy::Http::ResponseHeaderMap& headers, bool) {
+  if (isUpstreamResponse()) {
+    // Do not translate any response sent from the upstream.
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (Envoy::Grpc::Common::hasGrpcContentType(headers)) {
+    is_grpc_response_ = true;
+
+    if (isEspv2FilterError()) {
+      // gRPC error is already in filter state. No need to translate body, as it
+      // is not used for gRPC errors. Hence, nothing to do.
+      return FilterHeadersStatus::Continue;
```

Comment:
need to return StopIteration, otherwise it will be too late to modify response headers in encodeData()

Actually,  by reading envoy sendLocalReply() code,   if is_grpc,  it will not call encodeData() since it is trailer-only.